gnovm/pkg/gnolang/type_check.go

```diff
@@ -215,6 +215,145 @@ func assertAssignableTo(n Node, xt, dt Type, autoNative bool) {
+						case fv.Name == "cap":
+							assertValidConstValue(store, last, currExpr.Args[0], currExpr)
+							break Main
+						}
+					}
+				}
+			}
+
+			switch {
+			case len(tup.Elts) == 0:
+				panic(fmt.Sprintf("%s (no value) used as value", currExpr.String()))
+			case len(tup.Elts) == 1:
+				panic(fmt.Sprintf("%s (value of type %s) is not constant", currExpr.String(), tup.Elts[0]))
+			default:
+				panic(fmt.Sprintf("multiple-value %s (value of type %s) in single-value context", currExpr.String(), tup.Elts))
+			}
+		case *TypeType:
+			for _, arg := range currExpr.Args {
+				assertValidConstValue(store, last, arg, currExpr)
+			}
+		case *NativeType:
+			// Todo: should add a test after the fix of https://github.com/gnolang/gno/issues/3006
+			ty := evalStaticType(store, last, currExpr.Func)
+			panic(fmt.Sprintf("%s (variable of type %s) is not constant", currExpr.String(), ty))
+		default:
+			panic(fmt.Sprintf(
+				"unexpected func type %v (%v)",
+				ift, reflect.TypeOf(ift)))
+		}
+	case *BinaryExpr:
+		assertValidConstValue(store, last, currExpr.Left, parentExpr)
+		assertValidConstValue(store, last, currExpr.Right, parentExpr)
+	case *SelectorExpr:
+		xt := evalStaticTypeOf(store, last, currExpr.X)
+		switch xt := xt.(type) {
+		case *PackageType:
+			var pv *PackageValue
+			if cx, ok := currExpr.X.(*ConstExpr); ok {
+				// NOTE: *Machine.TestMemPackage() needs this
+				// to pass in an imported package as *ConstEzpr.
+				pv = cx.V.(*PackageValue)
+			} else {
+				// otherwise, packages can only be referred to by
+				// *NameExprs, and cannot be copied.
+				pvc := evalConst(store, last, currExpr.X)
+				pv_, ok := pvc.V.(*PackageValue)
+				if !ok {
+					panic(fmt.Sprintf(
+						"missing package in selector expr %s",
+						currExpr.String()))
+				}
+				pv = pv_
+			}
+			if pv.GetBlock(store).Source.GetIsConst(store, currExpr.Sel) {
+				break Main
+			}
+
+			tt := pv.GetBlock(store).Source.GetStaticTypeOf(store, currExpr.Sel)
+			panic(fmt.Sprintf("%s (variable of type %s) is not constant", currExpr.String(), tt))
+		case *PointerType, *DeclaredType, *StructType, *InterfaceType:
+			ty := evalStaticTypeOf(store, last, currExpr.X)
+			panic(fmt.Sprintf("%s (variable of type %s) is not constant", currExpr.String(), ty))
+		case *TypeType:
+			ty := evalStaticType(store, last, currExpr.X)
+			panic(fmt.Sprintf("%s (variable of type %s) is not constant", currExpr.String(), ty))
+		case *NativeType:
+			ty := evalStaticTypeOf(store, last, currExpr.X)
+			panic(fmt.Sprintf("%s (variable of type %s) is not constant", currExpr.String(), ty))
+		default:
+			panic(fmt.Sprintf(
+				"unexpected selector expression type %v",
+				reflect.TypeOf(xt)))
+		}
+
+	case *ArrayTypeExpr:
+		if parentExpr == nil {
+			ty := evalStaticTypeOf(store, last, currExpr)
+			panic(fmt.Sprintf("%s (variable of type %s) is not constant", currExpr.String(), ty))
+		}
+	case *ConstExpr:
+	case *BasicLitExpr:
+	case *CompositeLitExpr:
```

Comment:
can we just panic here?

Reply:
I deleted this cas I think we don't need it

Reply:
Why  you want we check the parent here ?

Reply:
Since we verify the expression type before calling this function, this case should not occur. Removed for array: https://github.com/gnolang/gno/pull/2828/commits/4da7934658d4e73d642cc3ceb554c4f6b0283f47

Reply:
will it be more straightforward that just check if `parentExpr` exist, and the type of `compositeLitExpr` is array type, otherwise panic?